atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,308 @@
-                retries += 1
-        return False
-
-    def lock(self, slot: int, prid: str):
-        logger.debug("Aquiring lock")
-        payload = json.dumps({"prid": prid, "locked": str(datetime.now())})
-        retries = 0
-        while retries < 3:
-            filename = os.path.join(ROOT_PATH, f'slot_{slot}.json')
-            _cmd = 'if grep -qw 0 {0} ; then echo {1} > {0} ;else exit 1; fi'.format(
-                    filename,
-                    quote(payload)
-                    )
-            cmd = 'flock {} -c {}; echo $?'.format(filename, quote(_cmd))
-            logger.debug(cmd)
-            ret = self._ssh_run_remote_command(cmd)
-            if ret == b'0\n':
-                logger.info('Locked')
-                return True
-            else:
-                logger.debug('Locking failed, retrying')
-                retries += 1
-        logger.info('Locking failed')
-        raise Exception('Locking failed')
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
+        return json_stdout
+
+    def available_slots(self) -> List[int]:
+        """
+        Returns list of available slots
+        """
+        logger.debug("Retrieve list of available slots on host %s", self.hostname)
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            available_slots = []
+            for slot_id in range(resource_info.get('max_slot_count', 1)):
+                cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+                slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+                if slot is None:  # Json load failed - file on remote host might be corrupted
+                    logger.warning("%s - slot %s - json load failed",
+                                   resource_info['hostname'], slot_id)
+                elif slot:  # Non-empty json file was loaded
+                    prid = slot.get('prid')
+                    logger.debug("%s - slot %s is occupied by %s",
+                                 resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available",
+                                 resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    logger.warning("%s - slot %s invalid state!",
+                                   resource_info['hostname'], slot_id)
```

Comment:
Not just undefined state, straight up impossible unless the `__bool__` method of slots returns non-deterministic results :smile:

Reply:
Yeah this should not ever happen :D .. I was not sure If I should add else or leave it be.

Reply:
I think the validation can be done earlier and continue the for-loop immediately when slot is None:

```
if slot is None:
    logger.warning()
    continue

self._validate_resource_lock(slot)

if slot:
    logger.debug("slot %s is occupied by %s")
else:
    logger.debug("slot is available.")
    available_slots.append(slot_id)
```

The data types are validated by the schema. Therefore, no "Undefined state" has to be cared about.

But, in either case, should the validation error be handled specifically? The `lock_resource` will fail if the validation fails. Then, will the pipeline run fail eventually? Or would it be good to log enough information for the invalid slot or the specific host and let the `lock_resource` locks other available resource as much as possible in order to keep the pipeline run?

Reply:
We discussed this a bit in one of the other threads - we want to log a warning and let the pipeline run try a different slot. Invalid slots could be fixed by the cleanup cronjob